Add Dockerfile and DockerCompose

MichaelTMS opened the pull request:
Running the massa node into a docker container.
build.sh script is used to build image and run container.
`cd docker && ./build.sh <Tag>`

Define variable values in docker-compose specification and run:
`docker-compose up -d`

damip commented:
Hello ! Thank you for your contribution.

Sorry for the delay, we are focusing on correcting the remaining bugs for testnet 15. We will review this PR after the delivery of testnet 15.

AurelienFT commented:
Hello Sorry @MichaelTMS for the delay. If you are still interested in working on this, I tried to launch `./build.sh 16.0` and I had the following error : 
```
❯ ./build.sh 16.0

Building massa-node docker image
Dockerfile:     /Users/aurelienfoucault/massa/docker/Dockerfile
docker context: /Users/aurelienfoucault/massa/docker
build date:     2022-11-22
image tag:      16.0

sed: 1: "/Users/xxx ...": command a expects \ followed by text
```

MichaelTMS commented:
> Hello Sorry @MichaelTMS for the delay. If you are still interested in working on this, I tried to launch `./build.sh 16.0` and I had the following error :
> 
> ```
> ❯ ./build.sh 16.0
> 
> Building massa-node docker image
> Dockerfile:     /Users/aurelienfoucault/massa/docker/Dockerfile
> docker context: /Users/aurelienfoucault/massa/docker
> build date:     2022-11-22
> image tag:      16.0
> 
> sed: 1: "/Users/xxx ...": command a expects \ followed by text
> ```

Hello. Which operating system is used when building the docker image?

AurelienFT commented:
> Hello. Which operating system is used when building the docker image?

I'm on MacOS 12 on a M1

MichaelTMS commented:
> > Hello. Which operating system is used when building the docker image?
> 
> I'm on MacOS 12 on a M1

Corrected the sed command for MacOS. Please try to build docker image

AurelienFT commented:
> Corrected the sed command for MacOS. Please try to build docker image

Still have the same error :/ : 
```
Building massa-node docker image
Dockerfile:     /Users/xxx/massa/docker/Dockerfile
docker context: /Users/xxx/massa/docker
build date:     2022-11-23
image tag:      16

sed: 1: "/Users/xxx ...": command a expects \ followed by text
```

MichaelTMS commented:
> > Corrected the sed command for MacOS. Please try to build docker image
> 
> Still have the same error :/ :
> 
> ```
> Building massa-node docker image
> Dockerfile:     /Users/xxx/massa/docker/Dockerfile
> docker context: /Users/xxx/massa/docker
> build date:     2022-11-23
> image tag:      16
> 
> sed: 1: "/Users/xxx ...": command a expects \ followed by text
> ```

Fixed the script. Added a command execution condition for different OS

AurelienFT commented:
I still have the same error. Will try to make it work as soon as I have a bit of time.

MichaelTMS commented:
> I still have the same error. Will try to make it work as soon as I have a bit of time.

Can you add the command "set -x" before "set -e" in build.sh script and send me output log, please

fsidhoum commented on docker/.dockerignore:
```diff
@@ -0,0 +1,5 @@
+.env
+.dockerignore
```
It is useless to explicitly ignore `.env` and `.dockerignore`: there are ignored by default.

By the way, you don't need to create such a file because in your Dockerfile, you copy explicitly the file you want to add to the Docker image.

fsidhoum commented on docker/.env:
```diff
@@ -0,0 +1,2 @@
+TAG=TEST.17.1
+COMPOSE_PROJECT_NAME=massa
```
Those kind of information should not be hardcoded but passed by the continuous integration through build arguments.

MichaelTMS replied:
The TAG parameter changes dynamically when executing docker/build.sh
if [[ "$(uname)" == "Darwin" ]]
then
    sed -i "" "s/^TAG=.*$/TAG=${IMAGE_TAG}/" "$DIR/.env"
else
    sed -i "s/^TAG=.*$/TAG=${IMAGE_TAG}/" "$DIR/.env"
fi

fsidhoum commented on docker/Dockerfile:
```diff
@@ -0,0 +1,46 @@
+# Builder Image
+FROM rust:1.63.0 AS builder
+
+ARG IMAGE_TAG
+ARG GIT_REPOSITORY
+
+# Set Working Directory is /app
+WORKDIR /app
+
+# Node project build
+RUN git clone --depth 1 --branch $IMAGE_TAG $GIT_REPOSITORY .
+RUN apt-get update && apt-get install -y build-essential libssl-dev libclang-dev
+RUN rustup toolchain install nightly-2022-11-14 && rustup default nightly-2022-11-14
+RUN cargo build --release --bin massa-node --bin massa-client 
```
Why to build the executables once again although there are already built during the release job ?

fsidhoum commented on docker/build.sh:
```diff
@@ -0,0 +1,57 @@
+done
+
+if [[ "$PUSH_FLAG" == "yes" ]]
+then
+    read -r -p "Enter DockerHub repository: " DOCKERHUB_REPO
+    IMAGE_NAME=$DOCKERHUB_REPO/$IMAGE_NAME:$IMAGE_TAG
+    read -r -p "Enter username: " DOCKERHUB_USERNAME
+    read -r -p "Enter password: " DOCKERHUB_PASSWORD
+else
+    IMAGE_NAME=$IMAGE_NAME:$IMAGE_TAG
+fi
+
+echo -e "\nBuilding massa-node docker image"
+echo -e "Build date: \t$BUILD_DATE"
+echo -e "Dockerfile: \t$DOCKERFILE"
+echo -e "Docker context: $DIR"
+echo -e "Docker Image: \t$IMAGE_NAME"
+echo -e "Version: \t$IMAGE_TAG\n"
+
+if [[ "$(uname)" == "Darwin" ]]
+then
+    sed -i "" "s/^TAG=.*$/TAG=${IMAGE_TAG}/" "$DIR/.env"
+else
+    sed -i "s/^TAG=.*$/TAG=${IMAGE_TAG}/" "$DIR/.env"
+fi
+
+docker build -f "$DOCKERFILE" "$DIR" \
+     --build-arg IMAGE_TAG="$IMAGE_TAG" \
+     --build-arg GIT_REPOSITORY="$GIT_REPOSITORY" \
+     --build-arg BUILD_DATE="$BUILD_DATE" \
+     --tag $IMAGE_NAME
+
+if [[ "$PUSH_FLAG" == "yes" ]]
+then
+    echo -e "\nSending docker image \"$IMAGE_NAME\" to DockerHub\n"
+    docker login -u $DOCKERHUB_USERNAME -p $DOCKERHUB_PASSWORD 2>/dev/null
+    docker push $IMAGE_NAME
+fi
+
+echo -e "\nThe build is complete\n"
```
Considering that the Docker image should be created during the process of the release, with a Github Action pipeline, this script is not usefull anymore.

fsidhoum commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,37 @@
+services:
+  massa-node:
+    container_name: massa-node
+    image: svetekllc/massa-node:${TAG}
```
Why using your personal namespace `svetekllc` ? Should probably be `massalabs`.

fsidhoum commented on docker/run_node.sh:
```diff
@@ -0,0 +1,68 @@
+    rm -rf $SOURCE_DIR/*
+fi
+
+## Massa node launch ###
+if [[ -f $PRIVKEY && -f $WALLET ]]
+then
+    echo "Node launch"
+    start_node
+else
+    echo "Node initialization"
+    init_node
+fi
```
Does this mean that we start the node only if the node is initialized ? If not, have I to relaunch the container, once the node is initialized ?

```suggestion
## Massa node launch ###
if [[ ! -f $PRIVKEY || ! -f $WALLET ]]
then
    echo "Node initialization"
    init_node
fi

echo "Node launch"
start_node
```

MichaelTMS replied:
@fsidhoum, I agree with the implementation, thank you very much.

fsidhoum commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,37 @@
+---
```
Not necessary.

fsidhoum reviewed (commented):
The creation and the publication of the Docker image should be integrated in the existing Github Action pipeline, and executed during the release.
